src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp

```diff
@@ -220,154 +299,99 @@ AttitudeEstimatorQ::Run()
+				_pos_acc.zero();
+				_vel_prev.zero();
+				_vel_prev_timestamp = 0;
 			}
 		}
+	}
+}
 
-		if (_local_position_sub.updated()) {
-			vehicle_local_position_s lpos;
-
-			if (_local_position_sub.copy(&lpos)) {
-
-				if (_param_att_acc_comp.get() && (hrt_elapsed_time(&lpos.timestamp) < 20_ms)
-				    && lpos.v_xy_valid && lpos.v_z_valid && (lpos.eph < 5.0f) && _inited) {
-
-					/* position data is actual */
-					const Vector3f vel(lpos.vx, lpos.vy, lpos.vz);
-
-					/* velocity updated */
-					if (_vel_prev_t != 0 && lpos.timestamp != _vel_prev_t) {
-						float vel_dt = (lpos.timestamp - _vel_prev_t) / 1e6f;
-						/* calculate acceleration in body frame */
-						_pos_acc = _q.rotateVectorInverse((vel - _vel_prev) / vel_dt);
-					}
-
-					_vel_prev_t = lpos.timestamp;
-					_vel_prev = vel;
-
-				} else {
-					/* position data is outdated, reset acceleration */
-					_pos_acc.zero();
-					_vel_prev.zero();
-					_vel_prev_t = 0;
+void AttitudeEstimatorQ::update_visual_odometry()
+{
+	if (_vehicle_visual_odometry_sub.updated()) {
+		vehicle_odometry_s vision;
+
+		if (_vehicle_visual_odometry_sub.update(&vision)) {
+			// validation check for vision attitude data
+			bool vision_att_valid = PX4_ISFINITE(vision.q[0])
+						&& (PX4_ISFINITE(vision.pose_covariance[vision.COVARIANCE_MATRIX_ROLL_VARIANCE]) ? sqrtf(fmaxf(
+								vision.pose_covariance[vision.COVARIANCE_MATRIX_ROLL_VARIANCE],
+								fmaxf(vision.pose_covariance[vision.COVARIANCE_MATRIX_PITCH_VARIANCE],
+										vision.pose_covariance[vision.COVARIANCE_MATRIX_YAW_VARIANCE]))) <= _eo_max_std_dev : true);
+
+			if (vision_att_valid) {
+				Dcmf Rvis = Quatf(vision.q);
+				Vector3f v(1.0f, 0.0f, 0.4f);
+
+				// Rvis is Rwr (robot respect to world) while v is respect to world.
+				// Hence Rvis must be transposed having (Rwr)' * Vw
+				// Rrw * Vw = vn. This way we have consistency
+				_vision_hdg = Rvis.transpose() * v;
+
+				// vision external heading usage (ATT_EXT_HDG_M 1)
+				if (_param_att_ext_hdg_m.get() == 1) {
+					// Check for timeouts on data
+					_ext_hdg_good = vision.timestamp_sample > 0 && (hrt_elapsed_time(&vision.timestamp_sample) < 500000);
 				}
 			}
 		}
-
-		/* time from previous iteration */
-		hrt_abstime now = hrt_absolute_time();
-		const float dt = math::constrain((now - _last_time) / 1e6f, _dt_min, _dt_max);
-		_last_time = now;
-
-		if (update(dt)) {
-			vehicle_attitude_s att = {};
-			att.timestamp_sample = sensors.timestamp;
-			_q.copyTo(att.q);
-
-			/* the instance count is not used here */
-			att.timestamp = hrt_absolute_time();
-			_att_pub.publish(att);
-
-		}
 	}
 }
 
-void
-AttitudeEstimatorQ::update_parameters(bool force)
+void AttitudeEstimatorQ::update_parameters(bool force)
 {
 	// check for parameter updates
 	if (_parameter_update_sub.updated() || force) {
 		// clear update
 		parameter_update_s pupdate;
-		_parameter_update_sub.copy(&pupdate);
+		_parameter_update_sub.update(&pupdate);
```

Comment:
I would keep it as a forced copy() (consistent with the other usage) and easier to refactor out later on if we come up with something better.